### ISSUE TITLE
Avoid loading unused dataset fields on snapshot pages

### ISSUE THREAD

**Issue by nellh** [OpenNeuroOrg/openneuro#1281]
Some information is loaded but not displayed (mostly dataset draft content). We should save bytes transferred by avoiding the extra content loading on snapshot pages. This would help when users are bandwidth limited.

ds001734 is a good example.

**Comment by exbotanical** [comment]
At first glance, this issue appears to be a simple query fix, but upon closer inspection we see that the query's resulting document is deeply nested throughout the OpenNeuro app. 

`getDatasetPage`—the query in question—calls a number of fragments, among them `DRAFT_FRAGMENTS`, which by default queries the entire draft file-set despite its not being used. The fragment cannot simply be removed, as several components rely on the drafts data in order to render properly. 

The dataset page exists in two discrete states at the top level (or should): 
1. users who are either admin and/or have write permissions -> these users need both snapshot files and drafts data
2. users viewing a dataset, but without admin and/or write permissions for said dataset -> these users are negatively impacted in that while they do not see draft files (nor need to), they nevertheless wait for them to load in the background. 

If you check the `left-sidebar `component, you can see an example of how write permissions are being checked in order to determine whether or not to visually render the drafts (namely, in the sidebar). This behavior needs to be matched by the query.

The solution here probably calls for a dynamic query. `DRAFT_FRAGMENT` (or at least the files field thereof, and their sub-fields) needs to toggle true / false contingent on the current user’s permissions level. The problem here is the query itself provides the `level` field, causing a recursive issue. We have a useful custom method called `hasEditPermissions `that requires `dataset` as an argument…but `dataset` is the document returned from our query.

I tried query-chaining. This doesn’t fare well, as neither GraphQL nor Apollo offer much coverage for resolving a query outside of a component / render. Follows are other solutions I’ve tried:

- adding a @ skip / @ include (inverses, so either is fine although one has greater specificity) directive into `DRAFT_FRAGMENT`, then determining the boolean directive variable by:
    - conditionally interpolating the variable, using a HOC 
    - hardcoding two separate queries (one without drafts), passed into two incarnations of `datasetQueryHook` -> recursively pulling out `hasEditPermissions `didn’t work 
- using Apollo’s `compose()` method
- using Apollo’s `withApollo` method
- using the ApolloProvider HOC
- creating an ApolloClient instance and accessing it with the imperative `.query()` call
- using `compose` with Apollo’s` useQuery` Hooks
- query chaining
- adding new fields to the schema
- editing the actual resolver function to require permissions (‘admin’ || ‘rw’) to resolve ; else, null
- passing field arguments through permissions function server-side  

**Note**: If you hardcode the boolean directive and pass it as props to `DatasetQueryHook`, the directive does work. It will omit draft files if @ skip === true || @ include === false

**Note**: If a proper instance of ApolloClient could be rendered (not sure where the necessary config, url, etc params are located), an isolated query could be resolved outside of the component. Then, its resulting value could produce the argument for the directive. The app is built a bit differently than examples in the Apollo spec, so I'm not sure about the instance. 

**Note**:` user` and `userId` are exposed in the instance’s context. Perhaps this would be helpful if someone thinks of a resolver-related solution. 

Theory: I have a theoretical solution that you can see as the most recent commit in my [working branch.](https://github.com/OpenNeuroOrg/openneuro/tree/1281-fix-draft-query)
It chains `useQuery `such that the directive variable is sort of ‘self-generated’. Here’s a simple example of what I was aiming for here:

```
const SPECIES_QUERY = gql `
    query SpeciesQuery {
        pet {
            id
            species
        }
    }
`
const PET_INFO_QUERY = gql `
    query PetInfoQuery
    ($getSomeInfo: Boolean!) {
        pet {
            id
            name
            gender
            someInfo @include(if: $getSomeInfo)
        }
    }
`

const { data: {pet: { species } = {} } = {} } = useQuery(SPECIES_QUERY)
const getPetInfo = species === 'DOG' || species === 'CAT'
const variables = 
{ getSomeInfo }
const skip = species === undefined
const { data: {pet } = {} } =
useQuery(SPECIES_QUERY, {variables, 
skip }) 
```

in the example, `species` will be undefined until the first query complete, so the second query will be skipped until that point. Once `species` is defined, `skip` evals to false but the value for `getSomeInfo` becomes valid and will be passed into the second query. 

I've never used gql/apollo prior to this, so I could certainly be missing a possibility, especially on the apollo side. I read the graphql spec and didn't see much that could help here other than changing the resolver/schema (which, strangely, didn't work or even change any resolution).

**Comment by exbotanical** [comment]
And I think a tl;dr is only fair, here:

**tl;dr** `getDatasetPage` is overfetching (loading draft files) for all users and not just those with permissions level ‘admin’ and/or ‘rw’. This is slowing things down considerably.

Two solution routes IMO are either server-side, authentication of permissions level in the resolver for draft files, OR using either the `@ skip` or `@ include` directives in the query, passing a boolean value (the variable of which is -> has admin or write permissions == true).

I’ve enumerated what I tried, and include a theoretical solution at the end. It should work, if the field-nesting can just be spread properly.

**Comment by nellh** [comment]
@MatthewZito The draft files only need to be conditional on the route. I thought there might be some lurking bugs with draft state updates when edits are made with this approach but in testing it, this does not seem to be a problem since all draft edits are already conditional on the draft state being loaded.

I'll push my branch and we can go over it together to see if there's more needed here.
